File: sdk/web/scripts/check-sqlite3mc-package.js
#!/usr/bin/env node
/** Verify the encryption-enabled npm package without running lifecycle scripts. */
import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { basename, isAbsolute, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const web = resolve(fileURLToPath(new URL('..', import.meta.url)));
const root = resolve(web, '../..');
const required = new Set([
  'package.json', 'js/index.js', 'js/types/api-types.d.ts',
  'js/key-vault.js', 'js/types/key-vault.d.ts',
  'dist/stellar_private_payments_web.js', 'dist/stellar_private_payments_web_bg.wasm',
  'dist/workers/storage-worker-module.js', 'dist/workers/storage-worker-module_bg.wasm',
  'dist/licenses/SQLite3MC.txt', 'dist/licenses/sqlite-wasm-vfs-LICENSE.txt',
  'dist/circuits/NOTICE.txt', 'dist/circuits/source-bundle.tar.gz',
]);
const notice = readFileSync(join(root, 'vendor/sqlite3mc-NOTICE.txt'));
const vfsNotice = readFileSync(join(root, 'vendor/sqlite-wasm-vfs/LICENSE'));

function npmPack(...args) {
  const result = JSON.parse(execFileSync('npm', ['pack', '--ignore-scripts', '--json', ...args], {
    cwd: web, encoding: 'utf8', stdio: ['ignore', 'pipe', 'inherit'],
  }));
  const packages = Array.isArray(result) ? result : Object.values(result);
  if (packages.length !== 1 || !packages[0]?.filename || !Array.isArray(packages[0]?.files)) {
    throw new Error('expected exactly one npm package result');
  }
  return packages[0];
}

function verifyNames(names) {
  for (const name of required) if (!names.has(name)) throw new Error(`package is missing ${name}`);
  for (const name of names) {
    if (isAbsolute(name) || name.split('/').includes('..') || name.includes('test-sqlite3mc') || name.startsWith('scripts/')) {
      throw new Error(`unexpected package entry: ${name}`);
    }
  }
}

function verify(read) {
  if (!read('dist/licenses/SQLite3MC.txt').equals(notice)) throw new Error('SQLite3MC notice differs from reviewed notice');
  if (!read('dist/licenses/sqlite-wasm-vfs-LICENSE.txt').equals(vfsNotice)) throw new Error('VFS notice differs from reviewed notice');
  const bindings = read('dist/stellar_private_payments_web.js').toString();
  for (const method of ['static openEncrypted(', 'static openMigration(', 'migrationAction(']) {
    if (!bindings.includes(method)) throw new Error(`WASM bindings lack ${method}`);
  }
  if (!read('js/index.js').toString().includes('recoverMigrationSetup')) throw new Error('JavaScript facade lacks initialization recovery');
  for (const name of ['dist/stellar_private_payments_web_bg.wasm', 'dist/workers/storage-worker-module_bg.wasm']) {
    if (!read(name).subarray(0, 4).equals(Buffer.from([0, 97, 115, 109]))) throw new Error(`invalid WASM artifact: ${name}`);
  }
}

const outputIndex = process.argv.indexOf('--output-directory');
const dryRun = npmPack('--dry-run');
const names = new Set(dryRun.files.map(({ path }) => path));
verifyNames(names);
verify((name) => readFileSync(join(web, name)));
const report = { passed: true, mode: 'dry-run', files: names.size };

if (outputIndex !== -1) {
  const output = resolve(process.argv[outputIndex + 1] ?? '');
  if (!process.argv[outputIndex + 1]) throw new Error('--output-directory needs a path');
  mkdirSync(output, { recursive: true });
  const packed = npmPack('--pack-destination', output);
  if (basename(packed.filename) !== packed.filename) throw new Error('npm returned an invalid package filename');
  const archive = join(output, packed.filename);
  const entries = execFileSync('tar', ['-tzf', archive], { encoding: 'utf8' }).trim().split('\n');
  if (entries.some((name) => !name.startsWith('package/') || name.endsWith('/'))) throw new Error('unexpected npm archive member');
  const archiveNames = new Set(entries.map((name) => name.slice('package/'.length)));
  verifyNames(archiveNames);
  verify((name) => execFileSync('tar', ['-xOzf', archive, `package/${name}`], { maxBuffer: 16 * 1024 * 1024 }));
  Object.assign(report, {
    mode: 'archive', archive: packed.filename,
    sha256: createHash('sha256').update(readFileSync(archive)).digest('hex'),
  });
  writeFileSync(join(output, 'verification.json'), `${JSON.stringify(report, null, 2)}\n`);
}
console.log(JSON.stringify(report, null, 2));

File: sdk/web/scripts/check-sqlite3mc-package.py
#!/usr/bin/env python3
"""Verify the staged encryption-enabled npm package; optionally create and inspect its archive.

Default mode checks npm's dry-run file list without copying the large circuit
bundle. --output-directory builds a CI artifact, never publishes it, and checks
the actual tar members. Lifecycle scripts are always disabled.
"""
import argparse
import hashlib
import json
from pathlib import Path, PurePosixPath
import subprocess
import tarfile

REQUIRED = {
    "package.json", "js/index.js", "js/types/api-types.d.ts",
    "js/key-vault.js", "js/types/key-vault.d.ts",
    "dist/stellar_private_payments_web.js", "dist/stellar_private_payments_web_bg.wasm",
    "dist/workers/storage-worker-module.js", "dist/workers/storage-worker-module_bg.wasm",
    "dist/licenses/SQLite3MC.txt", "dist/circuits/NOTICE.txt", "dist/circuits/source-bundle.tar.gz",
}


def package_result(raw):
    result = json.loads(raw)
    # npm 11 returns an array; npm 12 keys the same records by package name.
    packages = list(result.values()) if isinstance(result, dict) else result
    if not isinstance(packages, list) or len(packages) != 1 or not isinstance(packages[0], dict):
        raise ValueError("expected exactly one npm package result")
    package = packages[0]
    if not {"filename", "files"} <= package.keys():
        raise ValueError("npm package result lacks its filename or file list")
    return package


def verify(names, read, notice):
    missing = REQUIRED - names
    if missing:
        raise ValueError(f"encrypted package is missing: {sorted(missing)}")
    for name in names:
        path = PurePosixPath(name)
        if path.is_absolute() or ".." in path.parts or "test-sqlite3mc" in name or name.startswith("scripts/"):
            raise ValueError(f"unexpected package entry: {name}")
    if read("dist/licenses/SQLite3MC.txt") != notice:
        raise ValueError("SQLite3MC distribution notice differs from the reviewed source notice")
    bindings = read("dist/stellar_private_payments_web.js").decode()
    for method in ["static openEncrypted(", "static openMigration(", "migrationAction("]:
        if method not in bindings:
            raise ValueError(f"WASM bindings lack {method}; package may contain a legacy plaintext-only build")
    if "recoverMigrationSetup" not in read("js/index.js").decode():
        raise ValueError("JavaScript facade lacks initialization recovery")
    for name in ["dist/stellar_private_payments_web_bg.wasm", "dist/workers/storage-worker-module_bg.wasm"]:
        if read(name)[:4] != b"\0asm":
            raise ValueError(f"invalid WASM artifact: {name}")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output-directory", type=Path)
    args = parser.parse_args()
    web = Path(__file__).resolve().parents[1]
    notice = (web.parents[1] / "vendor/sqlite3mc-NOTICE.txt").read_bytes()
    # Preflight before creating a large archive, including a stock-build rejection.
    verify({name for name in REQUIRED if (web / name).is_file()},
           lambda name: (web / name).read_bytes(), notice)
    listing = package_result(subprocess.check_output(
        ["npm", "pack", "--dry-run", "--ignore-scripts", "--json"], cwd=web, text=True,
    ))
    verify({item["path"] for item in listing["files"]}, lambda name: (web / name).read_bytes(), notice)
    report = {"passed": True, "mode": "dry-run", "files": len(listing["files"])}
    if args.output_directory:
        output = args.output_directory.resolve()
        output.mkdir(parents=True, exist_ok=True)
        packed = package_result(subprocess.check_output(
            ["npm", "pack", "--ignore-scripts", "--json", "--pack-destination", str(output)],
            cwd=web, text=True,
        ))
        filename = packed["filename"]
        if Path(filename).name != filename:
            raise ValueError("npm returned an invalid package filename")
        archive = output / filename
        with tarfile.open(archive, "r:gz") as tar:
            files = {}
            for member in tar.getmembers():
                if not member.isfile() or not member.name.startswith("package/"):
                    raise ValueError(f"unexpected archive member: {member.name}")
                name = member.name.removeprefix("package/")
                if name in files:
                    raise ValueError(f"duplicate archive member: {name}")
                files[name] = member
            def read(name):
                with tar.extractfile(files[name]) as source:
                    return source.read()
            verify(set(files), read, notice)
        with archive.open("rb") as source:
            digest = hashlib.file_digest(source, "sha256").hexdigest()
        report.update(mode="archive", archive=filename, sha256=digest)
        (output / "verification.json").write_text(json.dumps(report, indent=2) + "\n")
    print(json.dumps(report, indent=2))


if __name__ == "__main__":
    main()
